### ISSUE TITLE
Is it possible to use your algorithms without any side effects on the disks, and also providing a past context of previous trials?

### ISSUE THREAD

**Issue by guillaume-chevalier** [Epistimio/orion#335]
If so, your algorithms should be compatible with the [Neuraxle](https://github.com/Neuraxio/Neuraxle) framework and vice-versa. 

I'd like to know. 

Neuraxle is a framework to allow for complex machine learning pipelines to be built. It especially allows for defining proper hyperparameter spaces within each step of the pipeline, and also allow for clean serialization of the different algorithms used and nested in the pipelines. 

Here is an example of automatic hyperparameter optimization using Neuraxle: 
https://www.neuraxle.org/stable/examples/boston_housing_meta_optimization.html#sphx-glr-examples-boston-housing-meta-optimization-py

**Comment by bouthilx** [comment]
If you want to use the python API (currently on develop branch, will be in next release v0.1.8), you can avoid any writes to file system provided that you use the mongodb backend or EphemeralDB. You may need to serialize the EphemeralDB in Neuraxle however to ensure persistence of the DB.

If you want to try building an example to integrate usage of Oríon with Neuraxle, I'm happy to help.

**Comment by guillaume-chevalier** [comment]
Glad to see you'd be happy to help! I think it'd be good to have some compatibility between Orion and Neuraxle, as Neuraxle provides neat pipelines and at least a RandomSearch and a TPE, while Orion could provide more AutoML algorithms and more ways to save trials to databases and such. 

Neuraxle's TPE is currently being coded by @Eric2Hamel [here](https://github.com/Neuraxio/Neuraxle/pull/224).

## Why I ask for decoupling objects

I would like to find a way to make your library compatible, and this implies that we'll need to remove the absolute need for interactions with disks or databases. Or at least, your library would need to allow to decouple those things. Let me explain why. 

First, one thing that is annoying with every or most hyperparameter tuning libraries is their lack of flexibility. For instance, it's near-impossible to extremely hard to take Hyperopt and to adapt it because of the way it's built. Their builtin TPE suggestion algorithm can't be really used outside of Hyperopt, and Hyperopt has a bad management of the saved trials. It's handy, but inflexible when comes the time to do changes to the project and adapt things to particular needs, such as resuming and restarting a search. See for instance this [big and still-unresolved Hyperopt issue](https://github.com/hyperopt/hyperopt/issues/267) where I've had some implications. 

The same goes for most other AutoML librairies. One thing that's needed is to allow for decoupling the saving of trials from the actual AutoML loop and from the actual suggestion algorithms. 

## By example with a random search: how we want things to be decoupled

See for instance our [RandomSearch class](https://github.com/Neuraxio/Neuraxle/blob/d7bcb268294744301f65748056c598c79ffcf8a6/neuraxle/metaopt/auto_ml.py#L606). It inherits from what we call the [AutoMLSequentialWrapper](https://github.com/Neuraxio/Neuraxle/blob/d7bcb268294744301f65748056c598c79ffcf8a6/neuraxle/metaopt/auto_ml.py#L452) and it registers itself into the sequential wrapper in the constructor providing the random search algorithm as the algorithm to use. Here is the signature of the constructor of the AutoMLSequentialWrapper: 

```
    def __init__(
            self,
            wrapped: BaseStep,
            auto_ml_algorithm: AutoMLAlgorithm,
            hyperparams_repository: HyperparamsRepository = None,
            n_iters: int = 100,
            refit=True,
            cache_folder_when_no_handle=None
    ):
```

### Here is a description of the most important parameters of that AutoMLSequentialWrapper class: 

- `wrapped` is a pipeline (black box) to be optimized. This is like a scikit-learn object where this objects needs to define the `fit` and `transform` method. This object also have `get_hyperparams` and `get_hyperparams_space` for instance, as well as setters. It's very easy to code wrapper classes to adapt for algorithms that would use any library (e.g.: we're already compatible with TensorFlow, PyTorch, scikit-learn, and a few others). 
- `auto_ml_algorithm` is something that inherits from the [AutoMLAlgorithm](https://github.com/Neuraxio/Neuraxle/blob/d7bcb268294744301f65748056c598c79ffcf8a6/neuraxle/metaopt/auto_ml.py#L276) class. This could be a RandomSearch or a TPE. It's the strategy used. The strategy have only two methods depicted below the present list. 
- `hyperparams_repository` is an abstract class that defines some abstract methods for saving and loading trials. This means that if you use databases, or disks, or anything, you only need to inherit from that class to add the good methods and voilà, anything can be used within our `AutoMLSequentialWrapper`. 

What's important to notice here is that the `AutoMLAlgorithm` is decoupled from the AutoML loop and from the repository where trials are saved. At any time, it's possible to combine any repository with any AutoMLAlgorithm, because each time the AutoMLAlgorithm is called, all the previous trials are sent and it can always predict on-the-fly even if the trials were made months ago as long as we pass the history to the new call. It has no stored state per-se (but it could have) and always receive the new trials. The important thing is that it doesn't call the database (repository) itself, so it's detached from it. 

## Upcoming improvements

Note that now we have an `AutoMLSequentialWrapper`, but we plan in the future to have `AutoMLParallelWrapper` with the same signature (or almost) to do optimization in parallel, and also `HyperbandAutoMLWrapper` to do partial runs like Hyperband (but using any other hp suggestion algorithm such as for instance TPE within it, and any repository). 

___

# To sum up

On your side, it'd be very good to be able to do this with Orion for Neuraxle: 
1. Regarding `AutoMLAlgorithm`: be able to take Orion's hyperparameter suggestion algorithms and wrap them into such an `AutoMLAlgorithm` class to allow their usage in Neuraxle (and in any or most other AutoML librairies. This involves also writing converters of Neuraxle hyperparameter spaces to Orion spaces. 
2. Regarding `Repositories`: be able to take Orion's databases and trial savers to wrap them into an `HyperparamsRepository` Neuraxle class. 
3. Regarding the `AutoMLSequentialWrapper`: If you have different such wrappers, you could add them to a Neuraxle-compatible wrapper class. 

On our side in Neuraxle, we could also do those 3 points in reverse: 
1. Allow for our algorithms to be used with Orion
2. Allow for our repositories to be used with Orion
3. Allow for our wrapper to ne used within Orion

@bouthilx Do you think this is easily feasible to write such wrappers? I think that decoupling the objects from each other will allow Orion to be compatible with more libraries.

**Comment by bouthilx** [comment]
I share your pain concerning Hyperopt :joy:. I tried implementing an algorithm plugin of TPE for Oríon by wrapping Hyperopt's implementation but it was close to impossible. I implemented one using Optuna instead ([here](https://github.com/bouthilx/orion.algo.optuna)), but it was still cumbersome to support their own database structure (I could not test thoroughly yet, I don't recommend using it for now).

About decoupling, I think our framework API could satisfy your needs: https://orion.readthedocs.io/en/develop/user/api.html#framework-api. It's not implemented yet, only the sequential and service API have been implemented so far, but we plan to release it with v0.2.0 (hopefully a few weeks from now).

There would still be the compatibility of Hyperparameter space definition which would be cumbersome to support between Neuraxle and Oríon. If I may ask, is there features in Neuraxle's hyperparameter definition that would be missing if using the space definition of another library? The way I see it Oríon and Neuraxle are very complementary. We deliberately avoided to include any scheduling dependency or pipelines to keep Oríon purely about black-box optimization, and a library to organize computation pipelines would be very useful to leverage Oríon in complex experiments. To be honest I had a look at Neuraxle a while ago while implementing my own library to automate my complex computation pipelines on HPC. I now stopped its implementation and planning to try out [Dask](https://dask.org/) instead, but I may consider trying out Neuraxle as well. It would be great to have usage examples in Oríon's documentation to use it with pipeline libraries.

In short, I'm wondering if the important value of Neuraxle is residing in the pipeline creation and execution and if duplication of efforts could be avoided by delegating Hyperparameter optimization specifics to Oríon. There is a need on our side for an integration with a pipeline framework and I feel we could work together. The way I see it, Oríon could be more tightly integrated within Neuraxle, this way relieving you from all development and maintenance of Hyperparameter optimization related code, and detailed examples/tutorials could be provided on how to use Oríon with Neuraxle, either in Oríon's documentation or in Neuraxle's documentation with Oríon's documentation clearly pointing to Neuraxle's doc. I know this is not what you had in mind when you made this issue, you clearly are seeking for a modular support for many hyperparameter optimization libraries (and so do we, with support for external plugins), but what do think of this proposition?

**Comment by guillaume-chevalier** [comment]
> I share your pain concerning Hyperopt :joy: .

Haha! Someone understands me! :wink: 

> There would still be the compatibility of Hyperparameter space definition which would be cumbersome to support between Neuraxle and Oríon. If I may ask, is there features in Neuraxle's hyperparameter definition that would be missing if using the space definition of another library?

The answer to this question depends on the way we see it an in which context. 

On one side, Neuraxle's hyperparams definition looks like the [scipy.stats](https://docs.scipy.org/doc/scipy/reference/stats.html) distributions API. So our hyperparams can be used within scikit-learn objects as well. We plan on supporting all scipy.stats' hyperparams too as our API is the same just with a few methods more or less on each hyperparam distribution. It wouldn't be hard to write adapters for other libraries. 

On the other side, our Pipelines can use nested hyperparam spaces and go from nested to not nested easily and such for the spaces and for the params too. Here are our spaces in the two versions: 

```
# Flat: 
    {
         "b__a__learning_rate": 7,
         "b__learning_rate": 9
     }
# Nested:  
     {
         "b": {
             "a": {
                 "learning_rate": 7
             },
             "learning_rate": 9
         }
     }
```

Our API can be summarized by reading [this example class](https://www.neuraxle.org/stable/hyperparameter_tuning.html#Creating-your-own-distributions) that has 3 functions. Note: we've also added the pdf and cdf function, and may support more functions of scipy.stats soon such as min, max, and mean, etc. 

You can learn more by skimming the code of [this file first](https://github.com/Neuraxio/Neuraxle/blob/master/neuraxle/hyperparams/space.py) and then [this file](https://github.com/Neuraxio/Neuraxle/blob/master/neuraxle/hyperparams/distributions.py). Or the [unit tests](https://github.com/Neuraxio/Neuraxle/tree/master/testing/hyperparams) are good as well.

> The way I see it Oríon and Neuraxle are very complementary.

Orion may be a library more for auto ml and Neuraxle for pipelines. Good. 

I think Neuraxle would still keep at least a TPE or something to do AutoML by default, and I wouldn't send that completely to another library unless I carefully review the licenses of all the dependencies of the other library which may be long as you seem to have a lot of dependencies. I don't accept to depend on "copyleft" licenses such as the GPL, for instance. 

It's interesting to note that I have myself separated my code in many different libraries to avoid having a lot of dependencies in the base `Neuraxle`. So if someone needs the TensorFlow capabilities or the PyTorch capabilities, then he can install add-ons and won't need everything from the start. 
- https://github.com/Neuraxio/Neuraxle
- https://github.com/Neuraxio/Neuraxle-TensorFlow
- https://github.com/Neuraxio/Neuraxle-PyTorch

> In short, I'm wondering if the important value of Neuraxle is residing in the pipeline creation and execution and if duplication of efforts could be avoided by delegating Hyperparameter optimization specifics to Oríon. 
> [...] 
> detailed examples/tutorials could be provided on how to use Oríon with Neuraxle, either in Oríon's documentation or in Neuraxle's documentation with Oríon's documentation clearly pointing to Neuraxle's doc
> [...]
> what do think of this proposition?

I think that Neuraxle can be good for Orion for sure, and I'm willing to help to make it more compatible with your stuff. It's just super cool if you do more AutoML than we provide by default. 

I would still at least keep some built-in capabilities for doing at least some AutoML in Neuraxle such as TPE and a few things that we've already done or are [currently finishing up](https://github.com/Neuraxio/Neuraxle/pull/279). E.g.: scikit-learn still has a Random Search and isn't an AutoML lib. So here we'd have TPE and you'd have more. 

**On adapting Orion to Neuraxle** 

And if it is the reverse, that I'd adapt Orion to fit Neuraxle, I or we could create Neuraxle-Orion for the add-ons. I'm still not sure if we'd adapt Neuraxle to Orion or Orion to Neuraxle, but it'd be something like that. I see my documentation site as a merge of all the projects, so I will soon add Neuraxle-TensorFlow and Neuraxle-PyTorch pages and examples (gallery) to Neuraxle's doc. It could be the same for Neuraxle-Orion.

**Comment by bouthilx** [comment]
### Search Space

We also support nested search space and all scipy.stats distributions. :) There is no support for conditional dimensions for the moment but it is part of the Road-Map, more long term however.

### Licences and dependencies

As far as I know there should be no issues with copyleft licenses. Oríon and Neuraxle have many dependencies in common actually.

We also share this view that dependencies should be kept minimal and this is why we keep algorithms that require specific dependencies outside of Oríon's core in independent plugins, like [algo.orion.skopt](https://github.com/Epistimio/orion.algo.skopt).

### Adapting Oríon or Neuraxle.

If need be and if this would improve Oríon in general, some modifications could be done in the core, otherwise an adaptation of Oríon for Neuraxle would probably be done through plugins, again to keep dependencies of Oríon minimal and vice-versa. Depending on what we are aiming for, that could be done through the storage protocol (for which we don't officially support plugins but all the structure is there to support it), or through algorithm plugins if it's just to use Neuraxle's algorithms within Oríon. But as I understand it the main advantage for both would be to provide support for algorithms of Oríon within Neuraxle.

A more tricky but probably more promising collaboration would be on refactoring the internal workflow of Oríon's workers to support other libraries like Neuraxle. There is already a will to do it for [cqueue](https://github.com/Delaunay/cqueue), we could look at Neuraxle at the same time. There isn't much bandwidth here to proceed with this at the moment however, so I cannot commit on when we would invest significant time on this.

### Nitpicking on terminology (sorry :sweat_smile: )

AutoML is not Hyperparameter optimization, it is a specific application of it. :wink: AutoML is specifically using algorithms for hyperparameter optimization to select best learning algorthims for a given task (in some sense we can say that AutoML is a problem of hyperparameter optimization where the search space include selection of learning algorithm type [random forest, neural network, svm, ...]). This means for instance that TPE is an algorithm for hyperparameter optimization, which can also be used for AutoML. In Oríon we focus on hyperparameter optimization in general. There is no plan for now to provide examples for AutoML, but that may be something interesting to do. Probably wrapping [auto-sklearn](https://github.com/automl/auto-sklearn) would be the first good step for this.

**Comment by guillaume-chevalier** [comment]
@bouthilx 

> AutoML terminology

Thanks for telling me haha

> There is no support for conditional dimensions for the moment but it is part of the Road-Map, more long term however.

We're working on making some dimensions "nullable" : that is, if they are disabled momentarily, they take some predefined default value like zero. 

> Adapting Oríon or Neuraxle

So you'd like to adapt Neuraxle to be useable within Orion's workers. Concretely, how would you see this done?

**Comment by bouthilx** [comment]
The discussion have been continued offline. I suggest we close this issue and open new ones depending on the work we decide to do.
